docs/architecture/adr-004-coroutine-effect-system.md

```diff
@@ -0,0 +1,303 @@
+async fn main() {
+    let (tx_queue, rx_queue) = tokio::mpsc::channel(16);
+
+    let network_service = NetworkService::new();
+
+    let params = ConsensusParams::new(/* ... */);
+    let mut state = ConsensusState::new(MyContext, params);
+
+    tokio::select! {
+        network_event = network_service.recv_msg() => {
+            match network_event {
+                NetworkEvent::Vote(vote) => {
+                    process_input(Input::Vote(vote), &mut state, &metrics, &network_service, &tx_queue)
+                }
+                // ...
+            }
+        },
+
+        input = rx_queue.recv() => {
+            process_input(input, &mut state, &metrics, &tx_queue)
+        }
+    }
+}
+
+
+// Function to process an input for consensus
+pub async fn process_input(
+   &mut self,
+   input: Input<MyContext>,
+   state: &mut ConsensusState<MyContext>,
+   metrics: &Metrics,
+   network_service: &NetworkService,
+   input_queue: &Sender<Input<MyContext>>,
+) -> Result<(), ConsensusError<MyContext> {
+    // Call the process! macro with our external effect handler
+    process!(
+        input: input,
+        state: state,
+        metrics: metrics,
+        with: effect => handle_effect(effect, input_queue)
+    )
+}
+
+// Method for handling effects
+async fn handle_effect(
+    effect: Effect<MyContext>
+    network_service: &NetworkService,
+    tx_queue: &Sender<Input<MyContext>>,
+) -> Result<Resume<MyContext>, Error> {
+    match effect {
+        Effect::SignVote(vote, r) => {
+            // Logic to sign a vote
+            let signed_vote = sign_vote(vote);
+
+            Ok(r.resume_with(signed_vote))
+        },
+
+        Effect::VerifySignature(msg, pk, r) => {
+            // Logic to verify a signature
+            let is_valid = verify_signature(&msg, &pk);
+
+            Ok(r.resume_with(is_valid))
+        },
+
+        Effect::Publish(msg, r) => {
+            // Logic to publish a message over the network
+            network_service.publish(msg).await;
+
+            Ok(r.resume_with(()))
+        },
+
+        Effect::GetValue(height, round, timeout, r) => {
+            // Extract the timeout duration
+            let timeout_duration = get_timeout_duration(timeout);
+
+            // Spawn a task to build the value asynchronously
+            let tx_queue = tx_queue.clone();
+            tokio::spawn(async move {
+                // Build the value (collecting txs, executing, etc.)
+                let value = build_value(height, round, timeout_duration).await;
+
+                // Put the `ProposeValue` consensus input in a queue,
+                // for it to be processed by consensus at a later point.
+                if let Ok(value) = result {
+                    tx_queue.send(Input::ProposeValue(value));
+                }
+            });
+
+            // Resume consensus immediately
+            Ok(r.resume_with(()))
+        }
+
+        // Handle other effects...
+    }
+}
+```
+
+### Notes
```

Comment:
I would find it helpful to have a subsection here describing the common-case flow, ie the sequence of inputs/effects reaching to a decision.